crates/cadenza-lsp/src/core.rs

```diff
@@ -0,0 +1,106 @@
+//! Core LSP utilities shared between native and WASM implementations.
+
+use lsp_types::*;
+
+/// Convert cadenza parse errors to LSP diagnostics.
+pub fn parse_to_diagnostics(source: &str) -> Vec<Diagnostic> {
+    let parsed = cadenza_syntax::parse::parse(source);
+
+    parsed
+        .errors
+        .iter()
+        .map(|error| {
+            let start_pos = offset_to_position(source, error.span.start);
+            let end_pos = offset_to_position(source, error.span.end);
+
+            Diagnostic {
+                range: Range::new(start_pos, end_pos),
+                severity: Some(DiagnosticSeverity::ERROR),
+                code: None,
+                code_description: None,
+                source: Some("cadenza".to_string()),
+                message: error.message.clone(),
+                related_information: None,
+                tags: None,
+                data: None,
+            }
+        })
+        .collect()
+}
+
+/// Convert a byte offset to an LSP Position.
+pub fn offset_to_position(text: &str, offset: usize) -> Position {
```

Comment:
These functions are going to be incredibly expensive, since they are run quite a bit and all scale linearly with the size of the module